### PR TITLE
#3709 - fix am/pm date handling

### PR DESCRIPTION
**Pull request by tatiana-scandi**
**Issue:**
* Fixes https://github.com/scandipwa/scandipwa/issues/3709

**Problem:**
If we enter PM time (greater than 12) + `12h` time format is enabled, date transformations (before the record in DB is created) lead to data object with hours > 12 and `PM` setting stored in it. This object is used in later calculations and transformed to `next day, am hours`, which is inserted to `quote_item_option` table. It explains wrong time in cart. 

**Solution:**
Use `h` format instead of `H`, when creating object with date data
(1 letter in line 64 changed, the rest of the code copied from Magento catalog-graphql without changes (had to copy as target method was private)).

### CODE REVIEW

File: src/etc/graphql/di.xml [hunk 1]
```diff
@@ -10,6 +10,7 @@
  */
 -->
 <config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:ObjectManager/etc/config.xsd">
+    <preference for="Magento\Catalog\Model\Product\Option\Type\Date" type="ScandiPWA\CatalogGraphQl\Model\Product\Option\DateType" />
```
Comment (zans-laksa) [inline src/etc/graphql/di.xml:13]: Please follow the same directory structure when you override some class.

**Review by zans-laksa — CHANGES_REQUESTED** [review]
There is something not clear to me.
Is the format configurable? Can we alter it from somewhere, like from the admin panel or some config files and if we can, what will happen if we set 24h time format?

**Comment by tatiana-scandi** [comment]
> There is something not clear to me. Is the format configurable? Can we alter it from somewhere, like from the admin panel or some config files and if we can, what will happen if we set 24h time format?

This format SHOULD be configurable:
https://github.com/scandipwa/scandipwa/issues/3658
There are options in Magento that allow to switch between 12h/24h.

But ScandiPWA uses default html calendar, which doesn't allow to customise date format.
Apart from this, using default calendar causes lots of browser specific bugs + doesn't allow to customise placeholders (QA are not happy about this) + bugs related to different user OS locale settings. 
Which is why I am planning to replace html inputs of type `date`, `time`, `datetime-local` with this: https://reactdatepicker.com/ (I got no suggestions when asked about this in chat, so just selected the most popular option Google suggested). For now this task is postponed, as I was assigned to another project for several weeks.

As for Luma, it doesn't use default HTML inputs, it uses custom JQuery calendar, it's definitely not a good idea to reuse it in ScandiPWA.

**Comment by zans-laksa** [comment]
I'd suggest to postpone this task until scandipwa/scandipwa#3658 is implemented and solve them both at once.
It's not good if we hardcode 12h format (lowercase `h`) and forget about this.
It's easy to miss such bug. We already did it. :)
@carinadues what do you think?

**Comment by tatiana-scandi** [comment]
> I'd suggest to postpone this task until [scandipwa/scandipwa#3658](https://github.com/scandipwa/scandipwa/issues/3658) is implemented and solve them both at once. It's not good if we hardcode 12h format (lowercase `h`) and forget about this. It's easy to miss such bug. We already did it. :) @carinadues what do you think?

Actually this Magento issue will remain after calendar update on FE. Now correct datetime format is passed to the server (I suppose this format should be always the same, no matter what date format is used for Date and Time display on FE). Magento logic correctly detects that date received from FE is PM time, but decides to add up 12 hours before creating a record in DB. As for `h` instead of `H`, it is not my hardcode, it is mistake of someone who wrote logic of Magento module that is not used initially.
https://monosnap.com/file/rfEdqzRDLxjgY3pWS9LfqCxW8vOb70

**Review by zans-laksa — APPROVED** [review]

**Pull request merged**